core/src/Cuda/Kokkos_Cuda_Parallel.hpp

```diff
@@ -749,25 +769,35 @@ class ParallelFor< FunctorType
     : m_functor( arg_functor )
     , m_policy( arg_policy )
     , m_league_size( arg_policy.league_size() )
-    , m_team_size( 0 <= arg_policy.team_size() ? arg_policy.team_size() :
-        Kokkos::Impl::cuda_get_opt_block_size< ParallelFor, LaunchBounds >( arg_functor , arg_policy.vector_length(), arg_policy.team_scratch_size(0),arg_policy.thread_scratch_size(0) ) / arg_policy.vector_length() )
+    , m_team_size( arg_policy.team_size() )
     , m_vector_size( arg_policy.vector_length() )
-    , m_shmem_begin( sizeof(double) * ( m_team_size + 2 ) )
-    , m_shmem_size( arg_policy.scratch_size(0,m_team_size) + FunctorTeamShmemSize< FunctorType >::value( m_functor , m_team_size ) )
-    , m_scratch_ptr{NULL,NULL}
-    , m_scratch_size{arg_policy.scratch_size(0,m_team_size),arg_policy.scratch_size(1,m_team_size)}
     {
+      cudaFuncAttributes attr = CudaParallelLaunch< ParallelFor, LaunchBounds >::
+          get_cuda_func_attributes();
+      m_team_size = m_team_size>=0?m_team_size:Kokkos::Impl::cuda_get_opt_block_size< FunctorType, LaunchBounds>(
+        m_policy.space().impl_internal_space_instance(),
+        attr, m_functor , m_vector_size,
+        m_policy.team_scratch_size(0), m_policy.thread_scratch_size(0) )/m_vector_size;
+
+      m_shmem_begin = ( sizeof(double) * ( m_team_size + 2 ) );
+      m_shmem_size = ( m_policy.scratch_size(0,m_team_size) + FunctorTeamShmemSize< FunctorType >::value( m_functor , m_team_size ) );
+      m_scratch_size[0] = m_policy.scratch_size(0,m_team_size);
+      m_scratch_size[1] = m_policy.scratch_size(1,m_team_size);
+
       // Functor's reduce memory, team scan memory, and team shared memory depend upon team size.
-      m_scratch_ptr[1] = cuda_resize_scratch_space(static_cast<ptrdiff_t>(m_scratch_size[1])*static_cast<ptrdiff_t>(Cuda::concurrency()/(m_team_size*m_vector_size)));
+      m_scratch_ptr[0] = NULL;
+      m_scratch_ptr[1] = m_team_size<=0?NULL:cuda_resize_scratch_space(static_cast<ptrdiff_t>(m_scratch_size[1])*static_cast<ptrdiff_t>(Cuda::concurrency()/(m_team_size*m_vector_size)));
 
       const int shmem_size_total = m_shmem_begin + m_shmem_size ;
-      if ( CudaTraits::SharedMemoryCapacity < shmem_size_total ) {
+      if ( m_policy.space().impl_internal_space_instance()->m_maxShmemPerBlock < shmem_size_total ) {
+        printf("%i %i\n",m_policy.space().impl_internal_space_instance()->m_maxShmemPerBlock,shmem_size_total);
```

Comment:
@crtrott is this intended to stay as extra info with the throw? If so can some context to the output values be added?